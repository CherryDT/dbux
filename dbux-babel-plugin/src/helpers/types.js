/**
 * Helpers to generate partial AST's.
 */

import { parse } from '@babel/parser';
import { codeFrameColumns } from "@babel/code-frame";
import traverse from "@babel/traverse";

/**
 * Generates AST from source through @babel/parser
 * @param {*} source 
 */
export function parseSource(source) {
  let ast;
  try {
    source = `${source}`;
    ast = parse(source);
  } catch (err) {
    const loc = err.loc;
    if (loc) {
      err.message +=
        "\n" +
        codeFrameColumns(source, {
          start: {
            line: loc.line,
            column: loc.column + 1,
          },
        });
    }
    throw err;
  }

  const nodes = ast.program.body;
  nodes.forEach(n => traverse.removeProperties(n));
  return nodes;
}

// https://stackoverflow.com/questions/35925798/how-to-add-an-import-to-the-file-with-babel