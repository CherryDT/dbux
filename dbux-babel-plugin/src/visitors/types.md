* `MemberExpression`
  * `OptionalMemberExpression`
* `Identifier`

* `AssignmentExpression`
* `VariableDeclarator`
* `VariableDeclaration`
* `ClassPrivateProperty`
* `ClassProperty`

* `Function`

* `CallExpression`
* `OptionalCallExpression`
* `NewExpression`
* `Super`

* Arithmetic
  * `BinaryExpression`
  * `UnaryExpression`
  * `LogicalExpression`

* `ConditionalExpression`
* `SequenceExpression`

* Value-creating/-changing
  * `UpdateExpression`
  * `TemplateLiteral`
  * `ObjectExpression`

* return/yield
  * `ReturnStatement`
  * `YieldExpression`

* `AwaitExpression`

* `Decorator`


# Statements

* `Declaration`
* `BreakStatement`
* `ContinueStatement`
* `ThrowStatement`
* `ForStatement`
* `ForInStatement`
* `ForOfStatement`
* `DoWhileLoop`
* `WhileStatement`
* `IfStatement`
* `SwitchStatement`
* `SwitchCase`
* `TryStatement`
* `CatchClause`
* `ExpressionStatement`
