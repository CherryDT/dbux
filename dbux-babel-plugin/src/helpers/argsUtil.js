import EmptyArray from '@dbux/common/src/util/EmptyArray';


function makeSpreadableArgumentCfg(argPath) {
  return {
    isSpread: argPath.isSpreadElement()
  };
}

/**
 * Takes a an array of arguments to indicate which are `SpreadElement` and which not.
 * 
 * NOTE: This is used by `CallExpression`, `ArrayExpression`.
 */
export function makeSpreadableArgumentArrayCfg(argumentPaths) {
  return argumentPaths?.map(makeSpreadableArgumentCfg) || EmptyArray;
}

/**
 * Takes a an array of arguments to indicate which are `SpreadElement` and which not.
 * 
 * NOTE: This is used by `ObjectExpression`.
 */
export function makeSpreadableArgumentObjectCfg(propertyPaths) {
  return propertyPaths?.map((propPath) => ({ 
    key: propPath.node.key,
    isSpread: propPath.isSpreadElement()
  })) || EmptyArray;
}
