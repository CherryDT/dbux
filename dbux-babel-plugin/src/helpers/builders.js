/**
 * Helpers to generate partial AST's.
 */

import { parse } from '@babel/parser';
import { codeFrameColumns } from "@babel/code-frame";
import traverse from "@babel/traverse";
import * as t from "@babel/types";
import { template } from '@babel/core';

export function buildNamedExport(ids) {
  return t.exportNamedDeclaration(null, ids.map(id => t.exportSpecifier(id, id)))
}


const errHandlerTemplate = template(`
console.error(err);
throw err;
`);

export function buildTryFinally(tryNodes, finallyNodes) {
  if (tryNodes.length === 1 && t.isBlockStatement(tryNodes[0])) {
    tryNodes = tryNodes[0];
  }
  else {
    tryNodes = t.blockStatement(tryNodes);
  }

  const catchClause = null;
  // const catchClause = t.catchClause(
  //   t.identifier('err'),
  //   t.blockStatement(errHandlerTemplate({
  //   }))
  // );
  return t.tryStatement(tryNodes, catchClause, t.blockStatement(finallyNodes));
}

export function buildBlock(statements) {
  return t.blockStatement(Array.isArray(statements) ? statements : [statements]);
}

export function buildVarDecl(ids) {
  const kind = "let";
  const declarations = ids.map((id, i) => {
    return t.variableDeclarator(id, null);
  });
  return t.variableDeclaration(kind, declarations);
}

export function buildVarAssignments(ids, values) {
  const assignments = ids.map((id, i) => {
    return t.expressionStatement(t.assignmentExpression('=', id, values[i]));
  });
  return assignments;
}

/**
 * Wrap a block with a try/finally pair
 */
export function buildWrapTryFinally(tryNodes, finallyBody) {
  tryNodes = Array.isArray(tryNodes) ? tryNodes : [tryNodes];
  return buildTryFinally(tryNodes, finallyBody);
}

export function buildProgram(origPathOrNode, bodyNodes) {
  const newProgramNode = t.cloneNode(origPathOrNode.node || origPathOrNode);
  newProgramNode.body = bodyNodes;
  return newProgramNode;
}

/**
 * Build AST from source through @babel/parser
 * @param {String} source 
 * @return {SourceNode[]}
 */
export function buildSource(source) {
  let ast;
  try {
    source = `${source}`;
    ast = parse(source);
  } catch (err) {
    const loc = err.loc;
    if (loc) {
      err.message +=
        "\n" +
        codeFrameColumns(source, {
          start: {
            line: loc.line,
            column: loc.column + 1,
          },
        });
    }
    throw err;
  }

  const nodes = ast.program.body;
  nodes.forEach(n => traverse.removeProperties(n));
  return nodes;
}

// https://stackoverflow.com/questions/35925798/how-to-add-an-import-to-the-file-with-babel