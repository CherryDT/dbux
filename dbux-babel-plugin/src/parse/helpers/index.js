import Declaration from './Declaration';

export {
  Declaration
};