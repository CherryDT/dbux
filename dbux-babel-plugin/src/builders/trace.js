import template from '@babel/template';
import * as t from '@babel/types';
import { newLogger } from '@dbux/common/src/log/logger';
import { getPresentableString } from '../helpers/pathHelpers';
import { bindTemplate } from '../helpers/templateUtil';

// eslint-disable-next-line no-unused-vars
const { log, debug, warn, error: logError } = newLogger('builders/trace');

const Verbose = 2;

export const buildTraceId = bindTemplate(
  '%%traceId%% = %%makeTraceId%%(%%staticTraceId%%)',

  function buildTraceId(templ, state, traceIdVar, inProgramStaticTraceId) {
    // TODO: add custom trace data
    const { ids: { aliases: {
      makeTraceId
    } } } = state;

    return templ({
      makeTraceId,
      staticTraceId: t.numericLiteral(inProgramStaticTraceId),
      traceId: traceIdVar
    });
  }
);

export const buildTraceExpression = bindTemplate(
  '%%traceExpression%%(%%expr%%, %%tid%%)',
  function buildTraceExpression(templ, path, state, traceType) {
    // TODO: add custom trace data
    // const { scope } = path;
    const { ids: { aliases: {
      traceExpression
    } } } = state;

    const tid = buildTraceId(path, state, traceType);
    const expr = path.node;
    Verbose && debug('[te]', getPresentableString(path));

    // NOTE: templates only work on `Node`, not on `NodePath`, thus they lose all path-related information.
    
    // TODO: keep `path` data etc, if necessary - `onCopy(path, newPath);`

    return templ({
      traceExpression,
      expr,
      tid
    });
  }
);

export const buildTraceNoValue = function buildTraceNoValue(templ, path, state, traceType) {
  const { ids: { dbux } } = state;
  const traceId = state.traces.addTrace(path, traceType);
  // console.warn(`traces`, state.traces);
  return templ({
    dbux,
    traceId: t.numericLiteral(traceId)
  });
}.bind(null, template('%%dbux%%.t(%%traceId%%)'));

