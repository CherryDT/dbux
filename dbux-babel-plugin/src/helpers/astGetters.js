/**
 * For reference, much of AST manipulation is based on babel-types:
 *   Source: https://github.com/babel/babel/tree/master/packages/babel-types
 *   Node classifications: https://github.com/babel/babel/tree/master/packages/babel-types/src/validators/generated/index.js
 *   Docs: https://babeljs.io/docs/en/babel-types
 *   ASTExplorer.net: https://astexplorer.net/
 */

import groupBy from 'lodash/groupBy';
import * as t from '@babel/types';
import { buildSource } from './builders';
import {isDebug} from 'dbux-common/src/util/nodeUtil';


/**
 * Might return null if given node has no ids (e.g. anonymous function declaration)
 */
export function getAllIdsOfDeclaration(node) {
  if (node.declaration.declarations) {
    // special case: variable declarations
    //  see: https://babeljs.io/docs/en/babel-types#variabledeclaration
    return node.declaration.declarations.map(n => n.id);
  }
  return node.declaration.id && [node.declaration.id] || null;
}

/**
 * 
 */
export function getAllClassParents(path) {
  const classParents = [];
  let classParent = path;
  do {
    classParent = classParent.findParent(p => t.isClass(p.node));
    if (classParent) {
      classParents.push(classParent);
    }
  } while (classParent);
  return classParents;
}