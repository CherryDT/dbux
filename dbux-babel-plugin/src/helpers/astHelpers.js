/**
 * For reference, much of AST manipulation is based on babel-types:
 *   Source: https://github.com/babel/babel/tree/master/packages/babel-types
 *   Node classifications: https://github.com/babel/babel/tree/master/packages/babel-types/src/validators/generated/index.js
 *   Docs: https://babeljs.io/docs/en/babel-types
 *   ASTExplorer.net: https://astexplorer.net/
 */

import groupBy from 'lodash/groupBy';
import * as t from '@babel/types';
import { buildSource } from './builders';


/**
 * Might return null if given node has no ids (e.g. anonymous function declaration)
 */
export function getAllIdsOfDeclaration(node) {
  if (node.declaration.declarations) {
    // special case: variable declarations
    //  see: https://babeljs.io/docs/en/babel-types#variabledeclaration
    return node.declaration.declarations.map(n => n.id);
  }
  return node.declaration.id && [node.declaration.id] || null;
}