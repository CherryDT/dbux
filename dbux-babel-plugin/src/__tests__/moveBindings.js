import { runAllSnapshotTests } from '../testing/test-util';
import { buildSource, buildVarDecl, buildVarAssignments } from '../helpers/builders';
import { replaceProgramBody } from '../helpers/modification';
import * as t from '@babel/types';

const codes = [
  `
{
  var a = 1;
  const b = 2;
  let c = 3;
}
`
];

const visited = new Set();

const plugin = function ({ types: t }) {
  return {
    visitor: {
      Program(path) {
        if (visited.has(path)) return;
        visited.add(path);

        const bodyPath = path.get('body');
        const nodes = bodyPath[0].get('body').map(path => path.node);
        const ids = [];
        const inits = [];
        for (let i = 0; i < nodes.length; ++i) {
          const node = nodes[i];
          const decls = node.declarations;
          if (decls) {
            nodes.splice(i--, 1);  // remove
            ids.push(...decls.map(n => n.id));
            inits.push(...decls.map(n => n.init));
          }
        }

        const newNodes = [
          buildVarDecl(ids),
          t.blockStatement(buildVarAssignments(ids, inits))
        ];

        replaceProgramBody(path, newNodes);
      }
    }
  };
};


runAllSnapshotTests(codes, __filename, plugin, true);
