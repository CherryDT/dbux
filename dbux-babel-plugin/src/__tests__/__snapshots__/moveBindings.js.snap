// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin moveBindings.es5.js: moveBindings.es5.js 1`] = `
"
/* #################################################################################### */

{
  var a = 1;
  const b = 2;
  let c = 3;
}

      ↓ ↓ ↓ ↓ ↓ ↓

\\"use strict\\";

var a, b, c;
{
  a = 1;
  b = 2;
  c = 3;
}
"
`;

exports[`plugin moveBindings.esNext.js: moveBindings.esNext.js 1`] = `
"
/* #################################################################################### */

{
  var a = 1;
  const b = 2;
  let c = 3;
}

      ↓ ↓ ↓ ↓ ↓ ↓

let a, b, c;
{
  a = 1;
  b = 2;
  c = 3;
}
"
`;
